std/format.d

```diff
@@ -7538,3 +7553,782 @@ if (is(T == float) || is(T == double) || (is(T == real) && T.mant_dig == double.
+            result[dot_pos + ((f.precision > 0 || f.flHash) ? 1 : 0) + f.precision + 1] = '+';
+        }
+        else
+        {
+            int sign_pos = cast(int) (result.length - 6);
+            if (f.precision>0 || f.flHash)
+            {
+                int dot_pos = cast(int) (result.length - f.precision - 5);
+                result[dot_pos] = '.';
+                sign_pos = dot_pos - 2;
+            }
+
+            if (f.flZero)
+                sign_pos = 0;
+            else if (sign_pos>0)
+                result[0 .. sign_pos + (sgn.length == 0 ? 1 : 0)] = ' ';
+
+            if (sgn != "")
+                result[sign_pos] = sgn[0];
+
+            result[$ - 3] = '+';
+            result[$ - 4] = is_upper ? 'E' : 'e';
+        }
+
+        return result.idup;
+    }
+
+    // add leading 1 for normalized values or correct exponent for denormalied values
+    if (exp != 0)
+        mnt |= 1L << (T.mant_dig - 1);
+    else
+        exp = 1;
+    exp -= bias;
+
+    // estimate the number of bytes needed left and right of the decimal point
+    // the speed of the algorithm depends on being as accurate as possible with
+    // this estimate
+
+    // Default for the right side is the number of digits given by f.precision plus one for the dot
+    // plus 6 more for e+...
+    auto max_right = f.precision + 7;
+
+    // If the exponent is <= 0 there is only the sign and one digit left of the dot else
+    // we have to estimate the number of digits. The factor between exp, which is the number of
+    // digits in binary system and the searched number is log_2(10). We round this down to 3.32 to
+    // get a conservative estimate. We need to add 3, because of the sign, the fact, that the
+    // logarithm is one to small and because we need to round up instead of down, which to!int does.
+    // And then we might need one more digit in case of a rounding overflow.
+    auto max_left = exp>0 ? to!int(exp / 3.32) + 4 : 3;
+
+    // If the result is not left justified, we may need to add more digits here for getting the
+    // correct width.
+    if (!f.flDash)
+        max_left = max(max_left, f.width - max_right + max_left + 1);
+
+    // If the result is left justified, we may need to add more digits to the right. This strongly
+    // depends, on the exponent, see above. This time, we need to be conservative in the other direction
+    // for not missing a digit; therefore we round log_2(10) up to 3.33.
+    if (exp > 0 && f.flDash)
+        max_right = max(max_right, f.width - to!int(exp / 3.33) - 2);
+    else if (f.flDash)
+        max_right = max(max_right, f.width);
+
+
+    byte[] buffer = new byte[max_left + max_right];
+    size_t start = max_left;
+    size_t left = max_left;
+    size_t right = max_left;
+
+    int final_exp = 0;
+
+    enum roundType { ZERO, LOWER, FIVE, UPPER }
+    roundType next;
+
+    if (exp >= T.mant_dig)
```

Comment:
There seems to be a lot of similarity between this branch and the branch below, can you facts it a bit more. its rather hard to follow.

Reply:
Not sure if you want me to add more comments or to explain it here?